llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -1859,7 +1859,7 @@ bool AMDGPUDAGToDAGISel::SelectScratchSVAddr(SDNode *N, SDValue Addr,
       std::tie(SplitImmOffset, RemainderOffset)
         = TII->splitFlatOffset(COffsetVal, AMDGPUAS::PRIVATE_ADDRESS, true);
 
-      if (isUInt<32>(RemainderOffset)) {
+      if (RemainderOffset > 0 && isInt<32>(RemainderOffset)) {
```

Comment:
Also needs matching GlobalISel change? Is it testable?

Reply:
The GlobalISel does not have such logic, so no changes needed there. I am not quite sure do we really need a test for this. I guess this can only be tested with out-of-bounds access, as we cannot allocate too much scratch, right? I felt the change is not quite useful in practice. what do you think?

Reply:
I think it's good to get all the edge cases right, even if it's an unrealistically large allocation. We haven't sorted out how TLS should work, but it's possible we'll end up using scratch instructions to access it at weird offsets

Reply:
While we can write LLVM IR like: `%ptr = getelementptr inbounds i8, ptr addrspace(5) %arg, i64 u0x80000000` for a positive offset with 31bit being one, it will be lowered into `t4: i32 = add t2, Constant:i32<-2147483648>` as our private address is 32bit. Seems like we cannot make a test to show difference here.